Support `Sync` trait for `rpm::Error`

I would like to use a return type equal to `anyhow::Result<RPMFile>`  for a multi-threaded code and I face the following issue:

```
        let header = rpm::PackageMetadata::parse(&mut buf_reader).context("RPM parsing failed")?;
```

```
error[E0599]: the method `context` exists for enum `Result<PackageMetadata, Error>`, but its trait bounds were not satisfied
   --> src/main.rs:140:67
    |
140 |         let header = rpm::PackageMetadata::parse(&mut buf_reader).context("RPM parsing failed")?;
    |                                                                   ^^^^^^^ method cannot be called on `Result<PackageMetadata, Error>` due to unsatisfied trait bounds
    |
   ::: /home/marxin/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/result.rs:502:1
    |
502 | pub enum Result<T, E> {
    | --------------------- doesn't satisfy `_: Context<PackageMetadata, Error>`
    |
   ::: /home/marxin/Programming/rpm-rs/src/errors.rs:9:1
    |
9   | pub enum Error {
    | --------------
    | |
    | doesn't satisfy `rpm::Error: Sync`
    | doesn't satisfy `rpm::Error: anyhow::context::ext::StdError`
    | doesn't satisfy `rpm::Error: std::marker::Send`
    |
    = note: the following trait bounds were not satisfied:
            `rpm::Error: anyhow::context::ext::StdError`
            which is required by `Result<PackageMetadata, rpm::Error>: anyhow::Context<PackageMetadata, rpm::Error>`
            `rpm::Error: std::marker::Send`
            which is required by `Result<PackageMetadata, rpm::Error>: anyhow::Context<PackageMetadata, rpm::Error>`
            `rpm::Error: Sync`
            which is required by `Result<PackageMetadata, rpm::Error>: anyhow::Context<PackageMetadata, rpm::Error>`
```

Which might not be possible due to:
```
error[E0599]: the method `as_dyn_error` exists for reference `&String`, but its trait bounds were not satisfied
   --> src/errors.rs:69:9
```
which is caused by fact that the GPG errors use `dyn Error`.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -9,7 +9,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Changed
 
-- `Error` now implements `Send + Sync`.
+- `Error` now implements `Send + Sync` (therefore, `Result<Package, Error>` now implements `Send + Sync`).
 
 ## 0.13.0
 

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -7,6 +7,10 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## Unreleased
 
+### Changed
+
+- `Error` now implements `Send + Sync`.
+
 ## 0.13.0
 
 ### Breaking Changes

--- a/src/errors.rs
+++ b/src/errors.rs
@@ -1,4 +1,4 @@
-use std::io;
+use std::{io, str::Utf8Error};
 
 use thiserror::Error;
 
@@ -57,17 +57,26 @@ pub enum Error {
     NoSignatureFound,
 
     #[error("error creating signature: {0}")]
-    SignError(Box<dyn std::error::Error>),
+    SignError(#[source] pgp::errors::Error),
 
-    #[error("error parsing key - {details}. underlying error was: {source}")]
-    KeyLoadError {
-        source: Box<dyn std::error::Error>,
-        details: &'static str,
-    },
+    #[error("error parsing keys, failed to parse bytes as utf8 for ascii armored parsing")]
+    KeyLoadUtf8Error(
+        #[from]
+        #[source]
+        Utf8Error,
+    ),
+
+    #[error("errors parsing keys, failed to parse bytes as ascii armored key")]
+    KeyLoadSecretKeyError(
+        #[from]
+        #[source]
+        pgp::errors::Error,
+    ),
 
     #[error("error verifying signature with key {key_ref}: {source}")]
     VerificationError {
-        source: Box<dyn std::error::Error>,
+        #[source]
+        source: pgp::errors::Error,
         key_ref: String,
     },
 
@@ -110,3 +119,8 @@ impl From<TimestampError> for Error {
         Error::TimestampConv(error)
     }
 }
+
+const _: () = {
+    const fn assert_send_sync<T: Send + Sync>() {}
+    assert_send_sync::<Error>();
+};

--- a/src/errors.rs
+++ b/src/errors.rs
@@ -56,6 +56,7 @@ pub enum Error {
     #[error("signature packet not found in what is supposed to be a signature")]
     NoSignatureFound,
 
+    #[cfg(feature = "signature-pgp")]
     #[error("error creating signature: {0}")]
     SignError(#[source] pgp::errors::Error),
 
@@ -66,6 +67,7 @@ pub enum Error {
         Utf8Error,
     ),
 
+    #[cfg(feature = "signature-pgp")]
     #[error("errors parsing keys, failed to parse bytes as ascii armored key")]
     KeyLoadSecretKeyError(
         #[from]
@@ -73,6 +75,7 @@ pub enum Error {
         pgp::errors::Error,
     ),
 
+    #[cfg(feature = "signature-pgp")]
     #[error("error verifying signature with key {key_ref}: {source}")]
     VerificationError {
         #[source]

--- a/src/errors.rs
+++ b/src/errors.rs
@@ -123,6 +123,7 @@ impl From<TimestampError> for Error {
     }
 }
 
+// Assert at compile-time that Error implements Send and Sync.
 const _: () = {
     const fn assert_send_sync<T: Send + Sync>() {}
     assert_send_sync::<Error>();

--- a/src/rpm/signature/pgp.rs
+++ b/src/rpm/signature/pgp.rs
@@ -63,12 +63,11 @@ impl traits::Signing for Signer {
         let passwd_fn = || self.key_passphrase.clone().unwrap_or_default();
         let signature_packet = sig_cfg
             .sign(&self.secret_key, passwd_fn, data)
-            .map_err(|e| Error::SignError(Box::new(e)))?;
+            .map_err(Error::SignError)?;
 
         let mut signature_bytes = Vec::with_capacity(1024);
         let mut cursor = io::Cursor::new(&mut signature_bytes);
-        pgp::packet::write_packet(&mut cursor, &signature_packet)
-            .map_err(|e| Error::SignError(Box::new(e)))?;
+        pgp::packet::write_packet(&mut cursor, &signature_packet).map_err(Error::SignError)?;
 
         Ok(signature_bytes)
     }
@@ -82,19 +81,13 @@ impl Signer {
     /// load the private key for signing
     pub fn load_from_asc_bytes(input: &[u8]) -> Result<Self, Error> {
         // only asc loading is supported right now
-        let input = std::str::from_utf8(input).map_err(|e| Error::KeyLoadError {
-            source: Box::new(e),
-            details: "Failed to parse bytes as utf8 for ascii armored parsing",
-        })?;
+        let input = std::str::from_utf8(input).map_err(Error::KeyLoadUtf8Error)?;
         Self::load_from_asc(input)
     }
 
     pub fn load_from_asc(input: &str) -> Result<Self, Error> {
         let (secret_key, _) =
-            SignedSecretKey::from_string(input).map_err(|e| Error::KeyLoadError {
-                source: Box::new(e),
-                details: "Failed to parse bytes as ascii armored key",
-            })?;
+            SignedSecretKey::from_string(input).map_err(Error::KeyLoadSecretKeyError)?;
         match secret_key.algorithm() {
             PublicKeyAlgorithm::RSA => Ok(Self {
                 secret_key,
@@ -159,9 +152,9 @@ impl traits::Verifying for Verifier {
             log::trace!("Signature has issuer ref: {:?}", key_id);
 
             if self.public_key.key_id() == *key_id {
-                return signature.verify(&self.public_key, data).map_err(|e| {
+                return signature.verify(&self.public_key, data).map_err(|source| {
                     Error::VerificationError {
-                        source: Box::new(e),
+                        source,
                         key_ref: format!("{:?}", key_id),
                     }
                 });
@@ -193,10 +186,10 @@ impl traits::Verifying for Verifier {
                             );
                             return Ok(());
                         }
-                        Err(e) => {
+                        Err(source) => {
                             log::trace!("Subkey verification failed");
                             result = Err(Error::VerificationError {
-                                source: Box::new(e),
+                                source,
                                 key_ref: format!("{:?}", sub_key.key_id()),
                             })
                         }
@@ -216,8 +209,8 @@ impl traits::Verifying for Verifier {
             );
             signature
                 .verify(&self.public_key, data)
-                .map_err(|e| Error::VerificationError {
-                    source: Box::new(e),
+                .map_err(|source| Error::VerificationError {
+                    source,
                     key_ref: format!("{:?}", self.public_key.key_id()),
                 })
         }
@@ -231,19 +224,13 @@ impl traits::Verifying for Verifier {
 impl Verifier {
     pub fn load_from_asc_bytes(input: &[u8]) -> Result<Self, Error> {
         // only asc loading is supported right now
-        let input = std::str::from_utf8(input).map_err(|e| Error::KeyLoadError {
-            source: Box::new(e),
-            details: "Failed to parse bytes as utf8 for ascii armored parsing",
-        })?;
+        let input = std::str::from_utf8(input).map_err(Error::KeyLoadUtf8Error)?;
         Self::load_from_asc(input)
     }
 
     pub fn load_from_asc(input: &str) -> Result<Self, Error> {
         let (public_key, _) =
-            SignedPublicKey::from_string(input).map_err(|e| Error::KeyLoadError {
-                source: Box::new(e),
-                details: "Failed to parse bytes as ascii armored key",
-            })?;
+            SignedPublicKey::from_string(input).map_err(Error::KeyLoadSecretKeyError)?;
 
         match public_key.algorithm() {
             PublicKeyAlgorithm::RSA => Ok(Self {

